File: singleList.java
import java.util.Scanner;

// Struktur Node Mahasiswa
class Node {
    String nim, nama;
    Node next;

    Node(String nim, String nama) {
        this.nim = nim;
        this.nama = nama;
        this.next = null; // Awalnya tidak menunjuk ke mana pun
    }
}

class LinkedList {
    Node head;        // Penunjuk awal list
    int count = 0;    // NB: Variabel count untuk melacak jumlah data

    // 1. Insert at Beginning
    void insertBeginning(String nim, String nama) {
        Node newNode = new Node(nim, nama);
        newNode.next = head; // Node baru menunjuk ke head yang lama
        head = newNode;      // Head berpindah ke node baru
        count++;             // NB 1: Setiap insert, count +1
    }

    // 2. Insert at Given Position
    void insertAtPos(String nim, String nama, int pos) {
        if (pos < 1 || pos > count + 1) {
            System.out.println("Posisi tidak valid!");
            return;
        }
        if (pos == 1) {
            insertBeginning(nim, nama);
            return;
        }
        Node newNode = new Node(nim, nama);
        Node temp = head;
        // Cari node sebelum posisi target
        for (int i = 1; i < pos - 1; i++) temp = temp.next;
        
        newNode.next = temp.next;
        temp.next = newNode;
        count++; // NB 1: Setiap insert, count +1
    }

    // 3. Insert at End
    void insertEnd(String nim, String nama) {
        Node newNode = new Node(nim, nama);
        if (head == null) {
            head = newNode;
        } else {
            Node temp = head;
            while (temp.next != null) temp = temp.next; // Cari ujung list
            temp.next = newNode;
        }
        count++; // NB 1: Setiap insert, count +1
    }

    // 4. Delete from Beginning
    void deleteBeginning() {
        if (head == null) return;
        head = head.next; // Lepaskan node pertama
        count--;          // NB 2: Setiap delete, count -1
    }

    // 5. Delete Given Position
    void deleteAtPos(int pos) {
        if (pos < 1 || pos > count) {
            System.out.println("Posisi tidak valid!");
            return;
        }
        if (pos == 1) {
            deleteBeginning();
            return;
        }
        Node temp = head;
        for (int i = 1; i < pos - 1; i++) temp = temp.next;
        temp.next = temp.next.next; // Melompati node yang dihapus
        count--; // NB 2: Setiap delete, count -1
    }

    // 6. Delete from End
    void deleteEnd() {
        if (head == null) return;
        if (head.next == null) {
            head = null;
        } else {
            Node temp = head;
            while (temp.next.next != null) temp = temp.next;
            temp.next = null;
        }
        count--; // NB 2: Setiap delete, count -1
    }

    // 7. Delete First Occurrence (by NIM)
    void deleteByNim(String nim) {
        Node temp = head, prev = null;
        while (temp != null) {
            if (temp.nim.equals(nim)) {
                if (prev != null) prev.next = temp.next;
                else head = temp.next;
                count--; // NB 2: Setiap delete, count -1
                System.out.println("Data NIM " + nim + " berhasil dihapus.");
                return;
            }
            prev = temp;
            temp = temp.next;
        }
        System.out.println("NIM tidak ditemukan.");
    }

    // 8. Show Data
    void show() {
        Node temp = head;
        System.out.println("\n--- DATA MAHASISWA (Total: " + count + ") ---");
        while (temp != null) {
            System.out.print("[" + temp.nim + " | " + temp.nama + "] -> ");
            temp = temp.next;
        }
        System.out.println("null");
    }
}

public class singleList {
    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        Scanner sc = new Scanner(System.in);
        int menu;

        do {
            // Menampilkan jumlah data saat ini secara real-time di atas menu
            System.out.println("\n[ Jumlah Mahasiswa Terdaftar: " + list.count + " ]");
            System.out.println("--- MENU LINKED LIST MAHASISWA ---");
            System.out.println("1-3: Insert, 4-6: Delete, 7: Del by NIM, 8: Show, 9: Exit");
            System.out.print("Pilih Menu (1-9): ");
            
            menu = sc.nextInt();
            sc.nextLine(); // Consume newline

            switch(menu) {
                case 1:
                    System.out.print("NIM: "); String nim1 = sc.nextLine();
                    System.out.print("Nama: "); String nama1 = sc.nextLine();
                    list.insertBeginning(nim1, nama1);
                    break;
                case 2:
                    System.out.print("NIM: "); String nim2 = sc.nextLine();
                    System.out.print("Nama: "); String nama2 = sc.nextLine();
                    System.out.print("Posisi: "); int p2 = sc.nextInt();
                    list.insertAtPos(nim2, nama2, p2);
                    break;
                case 3:
                    System.out.print("NIM: "); String nim3 = sc.nextLine();
                    System.out.print("Nama: "); String nama3 = sc.nextLine();
                    list.insertEnd(nim3, nama3);
                    break;
                case 4: list.deleteBeginning(); break;
                case 5: 
                    System.out.print("Hapus posisi ke: "); 
                    list.deleteAtPos(sc.nextInt()); 
                    break;
                case 6: list.deleteEnd(); break;
                case 7: 
                    System.out.print("Masukkan NIM: "); 
                    list.deleteByNim(sc.nextLine()); 
                    break;
                case 8: list.show(); break;
            }
        } while (menu != 9);
        System.out.println("Program Selesai.");
    }
}

File: singleList.py
# Definisi struktur Node (Gerbong Data)
class Node:
    def __init__(self, nim, nama):
        self.nim = nim      # Menyimpan data NIM
        self.nama = nama    # Menyimpan data Nama
        self.next = None    # Pointer ke node berikutnya (default: None)

class LinkedList:
    def __init__(self):
        self.head = None    # Awal mula list (kosong)
        self.count = 0      # NB: Variabel pelacak jumlah data

    # 1. Insert at Beginning (Data di depan)
    def insert_beginning(self, nim, nama):
        new_node = Node(nim, nama)
        new_node.next = self.head  # Node baru menunjuk ke head yang lama
        self.head = new_node       # Head pindah ke node baru
        self.count += 1            # NB 1: Insert, count +1

    # 2. Insert at Given Position (Posisi 1 sampai count+1)
    def insert_at_pos(self, nim, nama, pos):
        # Validasi apakah posisi masuk akal
        if pos < 1 or pos > self.count + 1:
            print("Posisi tidak valid!")
            return
        if pos == 1:
            self.insert_beginning(nim, nama)
            return
        
        new_node = Node(nim, nama)
        temp = self.head
        # Berjalan ke node tepat sebelum posisi tujuan
        for _ in range(pos - 2):
            temp = temp.next
        new_node.next = temp.next
        temp.next = new_node
        self.count += 1            # NB 1: Insert, count +1

    # 3. Insert at End (Data di akhir)
    def insert_end(self, nim, nama):
        new_node = Node(nim, nama)
        if not self.head:          # Jika list kosong
            self.head = new_node
        else:
            temp = self.head
            while temp.next:       # Cari node terakhir yang next-nya None
                temp = temp.next
            temp.next = new_node
        self.count += 1            # NB 1: Insert, count +1

    # 4. Delete from Beginning (Hapus depan)
    def delete_beginning(self):
        if not self.head: 
            print("List kosong!")
            return
        self.head = self.head.next # Head pindah ke node setelahnya
        self.count -= 1            # NB 2: Delete, count -1

    # 5. Delete Given Position (Posisi 1 sampai count)
    def delete_at_pos(self, pos):
        if pos < 1 or pos > self.count:
            print("Posisi tidak valid!")
            return
        if pos == 1:
            self.delete_beginning()
            return
        
        temp = self.head
        for _ in range(pos - 2):   # Cari node sebelum posisi target
            temp = temp.next
        temp.next = temp.next.next # Putus hubungan dengan node target
        self.count -= 1            # NB 2: Delete, count -1

    # 6. Delete from End (Hapus terakhir)
    def delete_end(self):
        if not self.head: return
        if not self.head.next:     # Jika hanya ada satu node
            self.head = None
        else:
            temp = self.head
            while temp.next.next:  # Berhenti di node sebelum terakhir
                temp = temp.next
            temp.next = None
        self.count -= 1            # NB 2: Delete, count -1

    # 7. Delete First Occurrence (Berdasarkan NIM)
    def delete_by_nim(self, nim):
        temp = self.head
        prev = None
        while temp:
            if temp.nim == nim:
                if prev:
                    prev.next = temp.next # Hubungkan node sebelumnya ke sesudahnya
                else:
                    self.head = temp.next # Jika yang dihapus adalah head
                self.count -= 1    # NB 2: Delete, count -1
                print(f"Data {nim} berhasil dihapus.")
                return
            prev = temp
            temp = temp.next
        print("NIM tidak ditemukan.")

    # 8. Show Data
    def show(self):
        temp = self.head
        print(f"\n--- DATA MAHASISWA (Jumlah: {self.count}) ---")
        if not temp:
            print("List Kosong")
            return
        while temp:
            print(f"[{temp.nim} | {temp.nama}]", end=" -> ")
            temp = temp.next
        print("None")

# Fungsi Main / Menu
# ... (kode class Node dan LinkedList tetap sama) ...

ll = LinkedList()
while True:
    # Menampilkan jumlah data saat ini secara real-time di atas menu
    print(f"\n[ Jumlah Mahasiswa Terdaftar: {ll.count} ]") 
    print("--- MENU LINKED LIST MAHASISWA ---")
    print("1. Insert Beg  2. Insert Pos  3. Insert End")
    print("4. Del Beg     5. Del Pos     6. Del End")
    print("7. Del by NIM  8. Show Data   9. Exit")
    
    pilihan = input("Pilih Menu (1-9): ")
    # ... (sisanya tetap sama) ...
    
    if pilihan == '1':
        ll.insert_beginning(input("NIM: "), input("Nama: "))
    elif pilihan == '2':
        nim, nama = input("NIM: "), input("Nama: ")
        pos = int(input(f"Posisi (1-{ll.count+1}): "))
        ll.insert_at_pos(nim, nama, pos)
    elif pilihan == '3':
        ll.insert_end(input("NIM: "), input("Nama: "))
    elif pilihan == '4':
        ll.delete_beginning()
    elif pilihan == '5':
        pos = int(input(f"Posisi dihapus (1-{ll.count}): "))
        ll.delete_at_pos(pos)
    elif pilihan == '6':
        ll.delete_end()
    elif pilihan == '7':
        ll.delete_by_nim(input("Masukkan NIM yang ingin dihapus: "))
    elif pilihan == '8':
        ll.show()
    elif pilihan == '9':
        print("Program Selesai.")
        break